### DISCUSSION TITLE
Beta testing pyalarmdotcomajax v0.5.0

### DISCUSSION THREAD

**Discussion by elahd** [pyalarmdotcom/alarmdotcom#265] [General]
[Release Notes](https://github.com/pyalarmdotcom/pyalarmdotcomajax/releases/tag/v0.5.0-beta.3)

### Installation

1. Install [Python >= 3.10](https://www.python.org/downloads/).
2. In a terminal, install [pyalarmdotcomajax](https://github.com/pyalarmdotcom/pyalarmdotcomajax) v0.5.0-beta.3 via pip: `pip install pyalarmdotcomajax==0.5.0-beta.3`

### Test 2FA

Test that two-factor authentication is working by running either of the other two tests in this post. You'll be prompted for your OTP.

### Test Device Data Retrieval

Run `adc -u 'your username' -p 'your password' get` to pull data for all of your devices. adc will print out a list of all supported devices, their states, and some attributes:

![image](https://github.com/pyalarmdotcom/alarmdotcom/assets/466460/761d2c13-62a3-48e6-a0da-b924e68b7d77)

Categories without supported devices will be empty.

![image](https://github.com/pyalarmdotcom/alarmdotcom/assets/466460/db8f4b62-7f6d-47f5-bddc-d2f9aee83d68)

There's a known bug in which the "System" category shows up empty.

### Test Real-Time Device Updates

Run `adc -u 'your username' -p 'your password' stream`, then do things with your alarm system (open windows/doors, arm your system, etc.). You should see real-time notices of these events in your terminal.

![image](https://github.com/pyalarmdotcom/alarmdotcom/assets/466460/fbac9c8c-3f9d-4633-97e9-61fd31f63ecd)

Note that some events are deliberately ignored as they are not yet supported or are duplicative.

Quit stream mode pressing Ctrl + C (Windows/Linux) or Cmd + C (Mac).

### Posting Results

Confirmation of both successes and failures are helpful. If you run into a failure, rerun the command while adding `-d -v` after `adc`, like this: `adc -d -v -u 'your username' -p 'your password' get`. This should return additional data that will be helpful in fixing the issue.

Thank you!

**Comment by Steini1984** [comment]
Good job on the update! 

I dont want to turn on 2fa (yet), but I can confirm this works for me just with user and pass. Same devices show up as with the older plugin and it shows the same statuses. 

Three points:
1. I had to change from -p "your password"  to -p 'your password'. I guess it´s because I have some special characters in my password.
2. By real time-notices do you mean <3 sek? In my testing it is about the same as with the older plugin. By that i mean up to 30 sek and sometimes only getting the closing event when opening and closing a door.
3. Smoke detectors are unknown in both plugins so I guess that is out of scope here:
ATTRIBUTES: [Type: Smoke Detector] [STATE: Unknown] [BATTERY: Normal]

**Reply by elahd** [reply]
Thanks!

> I had to change from -p "your password" to -p 'your password'. I guess it´s because I have some special characters in my password.

That may be OS-specific. I changed the instructions above, anyway, as both " and ' work for me in Windows.

> By real time-notices do you mean <3 sek? In my testing it is about the same as with the older plugin. By that i mean up to 30 sek and sometimes only getting the closing event when opening and closing a door.

I mean that when you use adc's `watch` function, you see event reports in the terminal as they happen in real life (within a couple of seconds). `watch` is powered by push notifications (WebSockets), which is new in this version of pyalarmdotcomajax. Prior versions only updated via pull/poll.

> Smoke detectors are unknown in both plugins so I guess that is out of scope here:
> ATTRIBUTES: [Type: Smoke Detector] [STATE: Unknown] [BATTERY: Normal]

Run the adc `get` command with the `-v` flag. Under `====[ Sensor ]====`, find the object for the smoke detector and post it here.

**Reply by Steini1984** [reply]
Perfect, thanks: 

Here is the relevant data:

```json
{
      "id":"98046728-26",
      "type":"devices/sensor",
      "attributes":{
         "deviceType":5,
         "openClosedStatus":0,
         "state":0,
         "isFlexIO":false,
         "deviceRole":0,
         "displayStateText":"Status Unknown",
         "isMonitoringEnabled":false,
         "sensorNamingFormat":4,
         "supportsBypass":false,
         "supportsImmediateBypass":false,
         "isBypassed":false,
         "managedDeviceType":3,
         "hasState":false,
         "canBeRenamed":true,
         "canBeDeleted":false,
         "canAccessWebSettings":true,
         "canAccessAppSettings":false,
         "webSettings":400,
         "canAccessTroubleshootingWizard":false,
         "troubleshootingWizard":"None",
         "addDeviceResource":0,
         "canBeAssociatedToVideoDevice":false,
         "associatedCameraDeviceIds":{
            
         },
         "macAddress":"",
         "manufacturer":"",
         "isAssignedToCareReceiver":false,
         "isOAuth":false,
         "isZWave":false,
         "supportsCommandClassBasic":false,
         "isMalfunctioning":false,
         "canBeSaved":true,
         "canChangeDescription":true,
         "description":"Uppi reykur",
         "deviceModelId":114,
         "canConfirmStateChange":true,
         "canReceiveCommands":false,
         "remoteCommandsEnabled":true,
         "hasPermissionToChangeState":true,
         "deviceIcon":{
            "icon":218
         },
         "batteryLevelNull":"None",
         "lowBattery":false,
         "criticalBattery":false
      },
      "relationships":{
         "system":{
            "data":{
               "id":"8084569",
               "type":"systems/system"
            }
         },
         "stateInfo":{
            "data":"None"
         }
      }
```

**Reply by elahd** [reply]
It looks like Alarm.com itself is reporting the status as unknown. What do you see for the detector at https://www.alarm.com/web/Security/Sensors.aspx?

Also, did the `watch` command work for you?

**Comment by Steini1984** [comment]
Well it is not the plugin so I might need to check with my provider if this is expected behaviour of the smoke alarms. 
![image](https://github.com/pyalarmdotcom/alarmdotcom/assets/5666926/e02eb1cb-06d6-484b-bea0-3a76ebd0a14d)

Yes the watch command works, but not working in realtime for me. I did some tests now with three doors.
Door 1 - Open ~30 sec to register
Door 1 - Close - almost instant
Door 1 - Open again - Not registered (waited for more then a minute)
Door 1 - Close  - registered a couple of minutes later while testing door 2 

Door 2 - Open ~30 sec to register
Door 2 - Close ~10 sec to register
Door 2 - Open again - Not registered (waited for more then a minute)
Door 2 - Close  - registered a couple of minutes later while testing door 3

Door 3- Open ~30 sec to register
Door 3 - Close - almost instant
Waiting 5 mins
Door 3- Open ~30 sec to register
Door 3 - Close - almost instant

The system I´m checking is also connected to the current home assistant plugin if that makes any difference. In other words same user and pass on both plugins at the same time.

**Reply by elahd** [reply]
Weird with the smoke detectors. Post back when they tell you what the problem is -- I'm curious as to what causes that.

For the door sensors, it shouldn't make a difference that you're logged into two places at once. Can you try two additional tests, separately?

1. Run the watch command in debug and verbose modes to see if any more immediate notifications are being ignored. `adc -d -v -u 'your username' -p 'your password' watch`.
2. Log into the Alarm.com website. There should be a "Recent Activity" pane on your dashboard. See how long it takes for door open/close events to show up there.

**Comment by Steini1984** [comment]
1. Nothing more in the verbose that was being ignored
2. Exactly the same time as in the terminal (this test was 35 sek open, 10 sek close)

**Reply by TheWebMachine** [reply]
I think this is just latency on Alarm.com's end of things. Even push notifications via native apps (Alarm.com and ADT, in my case) and SMS alerts are sometimes delayed by 30-40 seconds. I haven't been able to determine if the delay is originating locally at the panel -delayed sending status change to Alarm.com, or on their backend - not processing the status change in a timely manner.

**Comment by pdobrien3** [comment]
I would really like to try this but don’t have a Linux test box. Any chance you can provide some more details on testing on a windows box?  I have installed python 3.10 but can’t seem to move on from there. Thanks. 

After opening python 3.10 and running `pip install pyalarmdotcomajax==0.5.0-beta.2 ` I get

```
File "<stdin>", line 1
    pip install pyalarmdotcomajax==0.5.0-beta.2
        ^^^^^^^
SyntaxError: invalid syntax
```

**Reply by Steini1984** [reply]
I guess you need to install pip -> https://pip.pypa.io/en/stable/installation/

**Reply by elahd** [reply]
@pdobrien3 This works fine on Windows. pip is included with Python, so you don't need to install it separately. It looks like you're trying to run pip from within the Python interpreter. You need to run it from the C:\ prompt.

1. Hold down the Windows key and press R to open a Run... prompt.
2. Type `cmd` and press Enter.
3. Run `pip install pyalarmdotcomajax==0.5.0-beta.2` to install pyalarmdotcomajax.
4. Run the adc commands...

**Reply by pdobrien3** [reply]
I can’t get past the 2FA.  Bunch of errors. I tried with 2 different accounts. Do I need to escape symbols in passwords?

**Comment by elahd** [comment]
You don’t need to escape the characters — just put in quotation marks. What
type of 2FA are you using? SMS? App based? What are the errors?

**Reply by pdobrien3** [reply]
sms.  i will try and get you the errors tomorrow

**Comment by pdobrien3** [comment]


**Reply by elahd** [reply]
Update to `v0.5.0-beta.3` and try again. Also note that the `watch` command is now `stream`.

**Reply by elahd** [reply]
Also, if enclosing username and password in single quotes doesn't work, try double quotes (or vice versa). You may run into trouble if you have quotation marks in your password.

**Comment by pdobrien3** [comment]
That worked. 2FA worked perfectly. I had to use double quotes. I had pretty much the same experience. A door open took about 20 seconds whereas the close event happened almost immediately. I only had time to try one door though. Will play more later.

**Comment by mbhforum** [comment]
Works for me with 2FA as well.  Is there a way to incorporate this into the HA integration? Or were you just looking for feedback on whether this module works?

**Reply by elahd** [reply]
Working on it. Just wanted third party verification that changes to the core library work generally and not just for me.

**Comment by pdobrien3** [comment]
I believe it is an effort to improve the current adc custom component for Home Assistant.

**Comment by elahd** [comment]
I just released v3.0.0-beta of the integration with support for real time "push" updates and a fix for 2FA. 

If you don't see this version, go to HACS, click on integrations, click on Alarm.com, click on the 3 dot menu at the top right of the screen, click on redownload, and turn "show beta versions". v3 will appear shortly after turning this setting on.

**Comment by mikesalz** [comment]
I installed 3.0.0-beta.2 a day or so ago.  Since then, the alarm takes several minutes to update in HA.  Sometimes it does not update at all.

**Comment by pdobrien3** [comment]
Was it initially very responsive?

**Reply by mikesalz** [reply]
Just to clarify, are you asking: Did it perform well immediately after installing the beta and then degrade later?

I'm not sure, to be honest.  I upgraded and verified that it connected to my ADT account, but did not test further from there.  I noticed a couple of nights in a row that after I armed the alarm, it took several minutes before it was recognized in HA.  This morning I intentionally tested it and did not seem to register in HA at all.  At that point I rolled back to 2.2.1 which has been working pretty well.

**Reply by pdobrien3** [reply]
Yea, if you get a minute (I am just curious). Try going back to beta-2 and see if things degrade over time.

https://github.com/pyalarmdotcom/alarmdotcom/issues/262#issuecomment-1561931378

**Reply by mikesalz** [reply]
I'll give it a try over the weekend and post an update
